Refactoring/303 rewrite oracle integration test (#304)
exasol/virtual-schemas: doc/development/developing-sql-dialect/integration_testing_with_containers.md

# Integration Testing with Containers

> Virtual Schema integration tests use `exasol-testcontainers` framework, which requires docker privileged mode to be available to run the tests.

Reviewer: ```suggestion
Virtual Schema integration tests use the `exasol-testcontainers` framework, which requires docker privileged mode to be available to run the tests.
```

## Preparing Integration Test

> 1. In order to run the integration test automated, add the test to the includes list of the `maven-failsafe-plugin` in the [pom file](../../../pom.xml).

Reviewer: ```suggestion
1. In order to run the automated integration test, add the test to the includes list of the `maven-failsafe-plugin` in the [pom file](../../../pom.xml).
```

2. Provide a JDBC driver JAR for the source database.
3. Add a new Integration Test class for you database

### Provide JDBC drivers for the Source Database

> The JDBC drivers are automatically deployed during the test. You have to create a directory for the jdbc driver under `src/test/resources/integration/driver`. 

Reviewer: ```suggestion
The JDBC drivers are automatically deployed during the test. You have to create a directory for the JDBC driver under `src/test/resources/integration/driver`. 
```

The folder contains the driver jar file(s) and a `settings.cfg` file (for any integration test except Exasol and Postgres).

> In order to connect to the source database from your integration test you also have to add the jdbc driver dependency to the [POM](../../../pom.xml) scope verify.

Reviewer: Add an explanation that we can only deliver drivers where the license allows redistribution. Users should understand, why it is so cumbersome to get the test running.

Author: done

You can start the integration tests as follows:

```bash
mvn clean package && mvn verify
```

> This will run all included into the `maven-failsafe-plugin` integration tests.

Reviewer: ```suggestion
This will run all tests included in the `maven-failsafe-plugin` integration test configuration.
```

Another way to run integration tests:

* Create a package of Virtual Schemas using `mvn package` command and run integration tests inside your IDE in the same way as unit tests.

## Executing Disabled Integration Tests

Some integration tests are not running automatically, but it is possible to execute them locally. For example, `OracleSqlDialectIT`.

> In order to start `OracleSqlDialectIT`:

Reviewer: Add empty line below.

Author: done

> * Download Oracle JDBC driver `ojdbc8.jar` and oracle instant client `instantclient-basic-linux.x64-12.1.0.2.0.zip` and temporary put them into `src/test/resources/integration/driver/oracle` directory.

Reviewer: Enumerate.

Author: Done

* Run the tests from an IDE or temporary add `OracleSqlDialectIT.java` into the `maven-failsafe-plugin` includes section and execute  `mvn verify` command.

> * Remove the driver and the instant client after the test. Do not upload them to the GitHub repository.

Reviewer: Better: add everything into `.gitignore` so that people can't accidentally check those drivers in!

Author: Good idea!

Author: done
